### PR TITLE
Add support to tm1621 display

### PR DESCRIPTION
**Pull request by Philippe12**
# What does this implement/fix?

App support for TM1621 display controler that used in Sonoff THR316D, THR320D, POWR316D or POWR320D
Thanks to arendst for it first code into [Tasmota](https://github.com/arendst/Tasmota)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2242

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
display:
  platform: tm1621
  id: tm1621_display
  cs_pin: GPIO17
  data_pin: GPIO5
  read_pin: GPIO23
  write_pin: GPIO18
  lambda: |-
    it.printf(0, "%.1f", id(my_sensor1).state);
    it.display_celsius(true);
    it.printf(1, "%.1f", id(my_sensor2).state);
    it.display_humidity(true);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/tm1621/display.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+TM1621Display = tm1621_ns.class_("TM1621Display", cg.PollingComponent)
+TM1621DisplayRef = TM1621Display.operator("ref")
+
+CONFIG_SCHEMA = display.BASIC_DISPLAY_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(TM1621Display),
+        cv.Required(CONF_CS_PIN): pins.gpio_output_pin_schema,
+        cv.Required(CONF_DATA_PIN): pins.gpio_output_pin_schema,
+        cv.Required("read_pin"): pins.gpio_output_pin_schema,
+        cv.Required("write_pin"): pins.gpio_output_pin_schema,
```
Comment (jesserockz) [inline esphome/components/tm1621/display.py]: Please make consts for these. They can just go above in this file.

File: esphome/components/tm1621/tm1621.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  delayMicroseconds(80);
+  this->read_pin_->digital_write(false);
+  delayMicroseconds(15);
+  this->write_pin_->digital_write(false);
+  delayMicroseconds(25);
+  this->data_pin_->digital_write(false);
+  delayMicroseconds(TM1621_PULSE_WIDTH);
+  this->data_pin_->digital_write(true);
+
+  for (unsigned char tm1621_command : TM1621_COMMANDS) {
```
Comment (jesserockz) [inline esphome/components/tm1621/tm1621.cpp]: Please replace all `unsigned char` with `uint8_t`

```suggestion
  for (uint8_t tm1621_command : TM1621_COMMANDS) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/tm1621/display.py (hunk 1 above)
Comment (Philippe12) [reply esphome/components/tm1621/display.py]: That done
https://github.com/Philippe12/esphome/commit/06918e5dc8bccb8079411d55c62a1f4958e1374a

File: esphome/components/tm1621/tm1621.cpp (hunk 2 above)
Comment (Philippe12) [reply esphome/components/tm1621/tm1621.cpp]: Done
https://github.com/Philippe12/esphome/commit/fb1161bef4f08d77d59a7eef40bd0e8b3f336973

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Please add the platform to an existing test yaml file

**Comment by Philippe12** [comment]
> Please add the platform to an existing test yaml file

That done

File: tests/test1.yaml [hunk 3]
```diff
@@ -2091,7 +2093,44 @@ switch:
...
+            0x00,
+            0xFF,
+            0x00,
+            0xFF,
+            0x00,
+            0xFF,
+            0x00,
+            0xFF,
+          ]
   - platform: template
```
Comment (jesserockz) [inline tests/test1.yaml]: Please dont change unrelated lines in the test files.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: tests/test1.yaml (hunk 3 above)
Comment (Philippe12) [reply tests/test1.yaml]: I revert some code modifications due to VisualCode option (automatic formating when saving).

Comment (jesserockz) [reply tests/test1.yaml]: I also use this, but I have it set to only format lines that are modified, not the whole file :wink:

**Review by jesserockz — APPROVED** [review]

**Pull request merged**